### PR TITLE
Refactor: unify package inspection logic

### PR DESCRIPTION
**Pull request by abn**
This change brings together logic repeated in various places to discover package information. As a side-effect of this various edge cases have been improved and we shouldhave one place to change the code for package inspection related logic.

In addition to the above, this alos cleans up some test code and also unify call of urlparse and download.

Note this requires python-poetry/core#22. In order for the builds to work, I have created a branch for core that works around a few non issues with the current poetry releases used in ci. Additionally, the FreeBSD tests are failing because of a symbolic link issue (unrelated),

# Pull Request Check List
- [x] Added **tests** for changed code.
- ~~Updated **documentation** for changed code.~~

### CODE REVIEW

**Comment by abn** [comment]
Builds are failing due to https://github.com/python-poetry/poetry/pull/2302

File: pyproject.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ classifiers = [
 # Requirements
 [tool.poetry.dependencies]
 python = "~2.7 || ^3.5"
-poetry-core = "^1.0.0a5"
+poetry-core = { git = "https://github.com/abn/core.git", branch = "file-uri-scheme-sdist" }
```
Comment (stephsamson) [inline pyproject.toml:26]: I assume this will be changed once [#22](https://github.com/python-poetry/core/pull/22) is resolved?

Comment (abn) [reply pyproject.toml:26]: Yes.

**Comment by abn** [comment]
@sdispater i have ensured that we only execute `python setup.py egg_info` for git/path dependencies as it was before.

File: poetry/inspection/info.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+from poetry.utils.env import VirtualEnv
+from poetry.utils.setup_reader import SetupReader
+from poetry.utils.toml_file import TomlFile
+
+
+logger = logging.getLogger(__name__)
+
+
+class PackageInfoError(ValueError):
+    def __init__(self, path):
```
Comment (stephsamson) [inline poetry/inspection/info.py:35]: What about type hinting for `path` here for consistency?
